File: scala/protein-translation/src/main/scala/ProteinTranslation.scala
object ProteinTranslation {

  private[this] val CODON_PROTEIN_DICT = Map(
    "AUG" -> "Methionine",
    "UUU" -> "Phenylalanine",
    "UUC" -> "Phenylalanine",
    "UUA" -> "Leucine",
    "UUG" -> "Leucine",
    "UCU" -> "Serine",
    "UCC" -> "Serine",
    "UCA" -> "Serine",
    "UCG" -> "Serine",
    "UAU" -> "Tyrosine",
    "UAC" -> "Tyrosine",
    "UGU" -> "Cysteine",
    "UGC" -> "Cysteine",
    "UGG" -> "Tryptophan",
    "UAA" -> "STOP",
    "UAG" -> "STOP",
    "UGA" -> "STOP"
  )

  def proteins(sequence: String): Seq[String] = {
    sequence.grouped(3).flatMap(e => {
      CODON_PROTEIN_DICT.get(e)
    }).takeWhile(e => !e.equals("STOP")).toSeq
  }
}

File: haskell/protein-translation/src/ProteinTranslation.hs
module ProteinTranslation(proteins) where

import Data.List.Split
import Data.Map as Map
import Control.Monad

proteins :: String -> Maybe [String]
proteins xs = sequence (takeWhile isNormalProtein (Prelude.map translate $ chunksOf 3 xs))
    where
        codon_protein_dict = Map.fromList
            [
                ("AUG", "Methionine"),
                ("UUU", "Phenylalanine"),
                ("UUC", "Phenylalanine"),
                ("UUA", "Leucine"),
                ("UUG", "Leucine"),
                ("UCU", "Serine"),
                ("UCC", "Serine"),
                ("UCA", "Serine"),
                ("UCG", "Serine"),
                ("UAU", "Tyrosine"),
                ("UAC", "Tyrosine"),
                ("UGU", "Cysteine"),
                ("UGC", "Cysteine"),
                ("UGG", "Tryptophan"),
                ("UAA", "STOP"),
                ("UAG", "STOP"),
                ("UGA", "STOP")
            ]
        translate cond = Map.lookup cond codon_protein_dict
        isNormalProtein protein = protein /= Just "STOP"
